backend/service/src/universal_codes/upsert.rs

```diff
@@ -0,0 +1,158 @@
+        ModifyUniversalCodeError::DgraphError(error)
+    }
+}
+
+#[derive(Clone)]
+pub struct UpsertUniversalCode {
+    pub code: String,
+    pub name: Option<String>,
+    pub description: Option<String>,
+    pub r#type: Option<String>,
+    pub category: Option<String>,
+}
+
+pub async fn upsert_entity(
+    // ctx: &ServiceContext,
+    client: &dgraph::DgraphClient,
+    updated_entity: UpsertUniversalCode,
+) -> Result<Entity, ModifyUniversalCodeError> {
+    // Validate
+    let _original_entity = validate(client, &updated_entity).await?;
+
+    // Generate
+    let entity_input = generate(updated_entity.clone())?;
+
+    let result = dgraph::upsert_entity(client, entity_input).await?;
+
+    if result.numUids == 0 {
+        return Err(ModifyUniversalCodeError::InternalError(
+            "Failed to update entity".to_string(),
+        ));
+    }
+    // Query to get the updated record
+
+    let updated_entity = entity_by_code(client, updated_entity.code.clone())
+        .await
+        .map_err(|e| {
+            ModifyUniversalCodeError::InternalError(format!(
+                "Failed to get updated entity by code: {}",
+                e.message()
+            ))
+        })?;
+
+    let updated_entity = match updated_entity {
+        Some(updated_entity) => updated_entity,
+        None => {
+            return Err(ModifyUniversalCodeError::InternalError(
+                "Unable to find updated entity".to_string(),
+            ))
+        }
+    };
+
+    // Audit logging
+    // match original_entity {
+    //     Some(original_entity) => {
+    //         audit_log_entry(
+    //             &ctx,
+    //             LogType::UniversalCodeUpdated,
+    //             Some(original_entity.code),
+    //             Utc::now().naive_utc(),
+    //         )?;
+    //     }
+    //     None => {
+    //         audit_log_entry(
+    //             &ctx,
+    //             LogType::UniversalCodeCreated,
+    //             Some(updated_entity.code.clone()),
+    //             Utc::now().naive_utc(),
+    //         )?;
+    //     }
+    // }
+
+    Ok(updated_entity)
+}
+
+pub fn generate(
+    updated_entity: UpsertUniversalCode,
+) -> Result<EntityInput, ModifyUniversalCodeError> {
+    Ok(EntityInput {
+        code: updated_entity.code.clone(),
+        name: updated_entity.name.clone(),
+        description: updated_entity.description.clone(),
+        r#type: updated_entity.r#type.clone(),
+        category: updated_entity.category.clone(),
+    })
+}
+
+pub async fn validate(
+    // ctx: &ServiceContext,
+    client: &dgraph::DgraphClient,
+    new_entity: &UpsertUniversalCode,
+) -> Result<Option<Entity>, ModifyUniversalCodeError> {
+    // Check if entity exists with the same description
+    if let Some(description) = new_entity.description.clone() {
+        let duplicated =
+            check_description_duplication(client, description, new_entity.code.clone()).await?;
+        if duplicated.is_some() {
```

Comment:
The upsert test is failing here for me? `duplicated` is always `Some`, but the data array inside should be empty when not a duplicate...

Reply:
Thanks good find!

Reply:
Fixed now (I think)